spec/unit/hanami/validations/predicates/schema/custom_spec.rb

```diff
@@ -120,6 +120,100 @@ def self.name
     end
   end
 
+  describe 'with custom predicate within predicates block followed by a predicates module' do
+    before do
+      @validator = Class.new do
+        include Hanami::Validations
+
+        def self.name
+          'Validator'
+        end
+
+        predicate :url?, message: 'must be an URL' do |current|
+          current.start_with?('http')
+        end
+
+        predicates(
+          Module.new do
+            include Hanami::Validations::Predicates
+            self.messages_path = "spec/support/fixtures/messages.yml"
+
+            predicate(:email?) do |current|
+              current.match(/@/)
+            end
+          end
+        )
+
+        validations do
+          required(:foo) { url? | email? }
+        end
+      end
+    end
+
+    describe 'with valid input' do
+      let(:input) { { foo: %w[http://hanamirb.org foo@mailinator.com].sample } }
```

Comment:
I don't think using `sample` in specs is a good idea. This introduces randomness to tests and when this feature breaks in the future (and for example start to only accept urls), it might be not immediately visible, because if url is randomly chosen, the test will pass. I think it's better to either have two attributes or create two specs (one for email, one for url).

Reply:
Addressed here: b780b305dd6ce15fbed49e079890ec53661a9b81